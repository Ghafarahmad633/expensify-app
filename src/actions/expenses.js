import uuid from "uuid";
// this is ADD_EXPENSE

export const addExpenses=(
    {
        desc='',
        note='',
        createdAt=0,
        amount=0
    }={}
)=>({
    type:"ADD_EXPENSE",
    expense:{
        id:uuid(),
        desc,
        note,
        createdAt,
        amount
    }

});

// this is REMOVE_EXPENSE
export const removeExpense=({id}={})=>({
    type: "REMOVE_EXPENSE",
    id

})
export const removeExpenseSelected=(id)=>({
    type:"DELETE_SELECTED",
    id
})
// This is editExpense
export const editExpense=(id,updates)=>({
    type:"EDIT_EXPENSE",
    id,
    updates

})
export const removeExpenseSelctedRange=({ids})=>({
    type:"REMOVE_EXPENSES_SELECTED_RANGE",
    ids

})