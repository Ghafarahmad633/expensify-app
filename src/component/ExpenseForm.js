import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'

const now=moment();
console.log(now.format('MMM,Do, YYYY'))
class ExpenseForm extends React.Component{
    constructor(props){
        super(props);
        this.state={
            decs: props.expense?props.expense.desc:"",
            note:props.expense?props.expense.note:"",
            amount:props.expense?(props.expense.amount/100).toString():"",
            createdAt:props.expense?moment(props.expense.createdAt):moment(),
            calendarFocused:false,
            error:undefined,
            buttonText:props.buttonText?props.buttonText:"Submit"
        }
    }
    onChangeDecs=(e)=>{
        const decs=e.target.value
        this.setState(()=>({decs}))
    };
    onNoteChange=(e)=>{

        const note=e.target.value
        this.setState(()=>({note}))

    };
    onAmountChange=(e)=>{

        const amount=e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}))

        }
    };
    onDateChange=(createdAt)=>{

       if(createdAt){
           this.setState(()=>({createdAt}))
       }

    };
    onFocusChange=({focused})=>{
        this.setState(()=>({calendarFocused: focused}))
    };
    onSubmit=(e)=>{
        e.preventDefault()
        if(!this.state.decs||!this.state.amount){
            this.setState(()=>({error: "Please Fill All Fields"}))
            console.log("Some Error")
        }
        else {
            this.setState(()=>({error: ''}))
            this.props.onSubmit(({
                desc:this.state.decs,
                amount:parseFloat(this.state.amount,10)*100,
                createdAt:this.state.createdAt.valueOf(),
                note:this.state.note

            }))
        }
    }
    render(){
        return(
            <div>
                <form onSubmit={this.onSubmit}>
                    <input
                        name='desc'
                        type='text'
                        placeholder='Decritpion'
                        onChange={this.onChangeDecs}
                        value={this.state.decs}
                    />
                    <input
                        type='text'
                        name='amount'
                        placeholder='Amount'
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day)=>false}
                    />
                    <textarea
                        name='note'
                        onChange={this.onNoteChange}
                        value={this.state.note}
                        placeholder='add your note'
                    >

                    </textarea>
                    <button type='submit'>{this.state.buttonText}</button>
                </form>
                {this.state.error!==undefined&&<p>Please Fill out the required fields</p>}
            </div>
        )
    }

}
export default ExpenseForm