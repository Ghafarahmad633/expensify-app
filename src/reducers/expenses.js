// Expense Reducer
const expendDefaultState=[];
export default  (state=expendDefaultState,action)=>{
    switch (action.type) {
        case "ADD_EXPENSE":
            return [
                ...state,
                action.expense
            ];
        case "REMOVE_EXPENSE":
            return state.filter(({id,note,desc})=>{
                console.log("Remove "+id)
                return id!==action.id
            });
        case "EDIT_EXPENSE":
            return state.map((expense)=>{
                if(expense.id===action.id){
                    return {
                        ...expense,
                        ...action.updates
                    }
                }
                else{
                    return expense
                }
            });
        case "REMOVE_EXPENSES_SELECTED_RANGE":
            return state.filter((exp)=>{
                return !action.ids.includes(exp.id)
            });
        case 'SET_EXPENSES':
            return action.expenses
        default:
            return state
    }
}