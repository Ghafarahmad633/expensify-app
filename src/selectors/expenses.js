import moment from 'moment'

// Get Visible

export default(expenses,{text,sortBy,startDate,endDate})=>{

    return expenses.filter((exepnse)=>{
        const createdAtMoment=moment(exepnse.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment,'day'):true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment,'day'):true

        const textmatch = exepnse.desc.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textmatch


    }).sort((a,b)=>{
        if(sortBy==='date'){
            return a.createdAt<b.createdAt?1:-1
        }
        else if(sortBy==='amount'){
            return a.amount<b.amount?1:-1
        }
    })

}