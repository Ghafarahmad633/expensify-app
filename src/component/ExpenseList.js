import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from '../component/ExpenseListItem'
import selcectExpense from '../selectors/expenses'
import {addItemToDelete,removeItemToDelete} from '../actions/deleteSelected'
export class ExpenseList extends React.Component{
    constructor(props){
        super(props)
    };
    state={
        selectAll:this.props.selectAllExpense===true?'checked':null
    }
    onSelectChange=(e)=>{
        if(e.target.checked){
            this.props.addItemToDelete({id:e.target.value})

        }
        else if(!e.target.checked){
            this.props.removeItemToDelete({id:e.target.value})
        }



    };
    render(){
        return(
            <div>
                <h1>Expense List</h1>
                {
                    this.props.expenses.length===0?(
                        <p>No Expenses List Found</p>
                    ):(
                        this.props.expenses.map((expenses)=>(
                            <ExpenseListItem
                                key={expenses.id}
                                {...expenses}
                                onSelectChange={this.onSelectChange}
                                selectAll={this.state.selectAll}
                            />
                        ))
                    )
                }

            </div>
        )
    }
}

const mapStatProps=(state)=>{
    return{
        expenses:selcectExpense(state.expense,state.filters),
        expenseToDelete:state.selectedToDeleted,
        selectAllExpense:state.selectAllExpense

    }
}
const mapDispatchToProps=(dispatch,props)=>{
    return{
        addItemToDelete:(data)=>dispatch(addItemToDelete(data)),
        removeItemToDelete:(data)=>dispatch(removeItemToDelete(data))
    }
}
export default connect(mapStatProps,mapDispatchToProps)(ExpenseList)