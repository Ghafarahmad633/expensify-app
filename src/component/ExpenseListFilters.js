import React from 'react'
import {connect} from 'react-redux'
import {setFiltertext} from "../actions/filters";
import {sortyByDate,sortyByAmount} from '../actions/filters'
import {setStartDate,setEndDate} from '../actions/filters'
import {removeExpenseSelctedRange} from '../actions/expenses'
import {refereshStateAfterDelete,setTrueSelectAll,setFalseSelectAll} from '../actions/deleteSelected'
import {DateRangePicker} from 'react-dates'
export class ExpenseListFilters extends React.Component {
    constructor(props){
        super(props)
    }
    state={
        calendarFocused:null,

    }
    onDatesChange=({startDate,endDate})=>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)

    };
    onFocusChange=(calendarFocused)=>{
        this.setState(()=>({calendarFocused}))
    };
    deletedAllSelected=()=>{
        this.props.removeExpenseSelctedRange({ids:this.props.selectedToDeleted})
        this.props.refereshStateAfterDelete()

    }
    selectAll=(e)=>{

        if(e.target.checked){
            this.props.dispatch(setTrueSelectAll())
        }
        if(!e.target.checked){
            this.props.dispatch(setFalseSelectAll())
        }
        console.log(this.props.selectAllExpense)
    }

    onTextChange=(e)=>{
        this.props.setFiltertext(e.target.value)
    }
    onSortChange=(e)=>{
        const filter_select=e.target.value
        if(filter_select==='date'){
            this.props.sortyByDate()
        }
        else if(filter_select==='amount'){
            this.props.sortyByAmount()
        }
    }


    render(){
        return(
            <div>

                <input type='text'
                       disabled={this.props.expenses.length<=0}
                       placeholder='Enter Value To Search'
                       value={this.props.filters.text} onChange={this.onTextChange}/>
                <input type='button' value='Delected Selected'
                       disabled={this.props.selectedToDeleted.length<=0}
                       onClick={this.deletedAllSelected}
                />
                <select
                    disabled={this.props.expenses.length<=0}
                    value={this.props.filters.sortyBy}
                    onChange={this.onSortChange}>
                    <option  value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                    showClearDates={true}
                    disabled={this.props.expenses.length<=0}
                />
                Select All <input
                type='checkbox'
                onClick={this.selectAll}
                value='selectAll'/>


            </div>
        )
    }

}

const mapStateProps=(state)=>{
    return{
        filters:state.filters,
        expenses:state.expense,
        selectedToDeleted: state.selectedToDeleted,
        selectAllExpense:state.selectAllExpense
    }
}
const mapDispatchToProps=(dispatch,props)=>{
    return{
        sortyByDate:()=>dispatch(sortyByDate()),
        sortyByAmount:()=>dispatch(sortyByAmount()),
        setFiltertext:(text)=>dispatch(setFiltertext(text)),
        setStartDate:(date)=>dispatch(setStartDate(date)),
        setEndDate:(date)=>dispatch(setEndDate(date)),
        removeExpenseSelctedRange:(data)=>dispatch(removeExpenseSelctedRange(data)),
        refereshStateAfterDelete:()=>refereshStateAfterDelete()

    }

}
export default connect(mapStateProps,mapDispatchToProps)(ExpenseListFilters)