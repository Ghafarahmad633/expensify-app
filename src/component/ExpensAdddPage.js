import React from "react";
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import {addExpenses} from '../actions/expenses'
export class ExpensAdddPage extends React.Component{
    onSubmit=(expenses)=>{
        this.props.addExpenses(expenses)
        this.props.history.push("/")
    }
    render(){
       return(
           <div>
               <h1>Add Expense</h1>
               <ExpenseForm onSubmit={this.onSubmit}/>
           </div>
       )
    }
}
const mapDispatchToProps=(dispatch)=>{
    return{
        addExpenses:(expenses)=>dispatch(addExpenses(expenses))
    }
}
export default connect(undefined,mapDispatchToProps)(ExpensAdddPage)