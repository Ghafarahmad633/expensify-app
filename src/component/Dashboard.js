import React from "react";

import ExpenseList from "./ExpenseList";
import ExpenseListFilters from './ExpenseListFilters'
class ExpensifyDashboardPage extends React.Component{

    render() {
        return (
            <div>
                <div>
                    <ExpenseListFilters/>
                    <ExpenseList />
                </div>
            </div>
        );
    }
}

export default ExpensifyDashboardPage