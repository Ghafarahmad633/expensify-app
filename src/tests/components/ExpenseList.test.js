import React from 'react'
import {ExpenseList} from '../../component/ExpenseList'
import {shallow} from 'enzyme'
import expenses from '../fixtures/expenses'
test('should render ExpenseList with expenses',()=>{
    const wraper =shallow(<ExpenseList expenses={expenses}/>)
    expect(wraper).toMatchSnapshot()
})
test('should render ExpenseList with no expenses',()=>{
    const wraper =shallow(<ExpenseList expenses={[]}/>)
    expect(wraper).toMatchSnapshot()
})