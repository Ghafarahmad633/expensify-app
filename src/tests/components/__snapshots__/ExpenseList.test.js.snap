// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <Connect(ExpenseListItem)
    amount={1000}
    createdAt={0}
    desc="Rent"
    id="1"
    key="1"
    note="This is home rent"
    onSelectChange={[Function]}
    selectAll={null}
  />
  <Connect(ExpenseListItem)
    amount={1800}
    createdAt={-345600000}
    desc="Gum"
    id="2"
    key="2"
    note="This is Gum expense"
    onSelectChange={[Function]}
    selectAll={null}
  />
  <Connect(ExpenseListItem)
    amount={3000}
    createdAt={345600000}
    desc="Fee"
    id="3"
    key="3"
    note="This is School Fee"
    onSelectChange={[Function]}
    selectAll={null}
  />
</div>
`;

exports[`should render ExpenseList with no expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    No Expenses List Found
  </p>
</div>
`;
